problemtools/verifyproblem.py

```diff
@@ -740,6 +741,296 @@ def check(self, args):
+        command = state[key]
+        name = os.path.basename(state['path'])
+        random_salt = str(state['random_salt'])
+
+        def err():
+            self.error('Invalid %s key for path %s in generators.yaml' % (key, state['path']))
+
+        if not isinstance(command, str):
+            return err()
+
+        seed = str(int(hashlib.sha512((random_salt + command).encode('utf-8')).hexdigest(), 16) % (2**31))
+
+        parts = shlex.split(command)
+        if not parts:
+            return err()
+
+        for i, part in enumerate(parts):
+            new = ''
+            for j, group in enumerate(part.split('{')):
+                if group.count('}') != (0 if j == 0 else 1):
+                    return err()
+                if j == 0:
+                    new += group
+                else:
+                    group, rest = group.split('}')
+                    if group.startswith('seed'):
+                        new += seed
+                    elif group == 'name':
+                        new += name
+                    else:
+                        return err()
+                    new += rest
+            parts[i] = new
+
+        program, arguments = parts[0], parts[1:]
+        if program not in self._generators:
+            self._generators[program] = program
+
+        return (program, arguments)
+
+    def _parse_testcase(self, data, state):
+        if state['input'] is None:
+            self.error('Path %s in generators.yaml must contain an input key' % state['path'])
+        for key in ['input', 'solution', 'visualizer']:
+            if state[key] is not None:
+                state[key] = self._parse_command(key, state)
+        if state['input'] is not None:
+            self._testcases.append(state)
+
+    def _parse_directory(self, data, state):
+        # TODO: Process includes
+
+        if 'testdata.yaml' in data:
+            content = data['testdata.yaml']
+            if content is None:
+                content = {}
+            self._testdata_yaml['%s/%s' % (state['path'], 'testdata.yaml')] = content
+
+        cases = data.get('data', {})
+        ordered = True
+        if not isinstance(cases, list):
+            ordered = False
+            cases = [cases]
+
+        case_counter = 0
+        case_format = '%%0%dd' % len(str(len(cases)))
+        for case in cases:
+            if not isinstance(case, dict):
+                self.error('Path %s/data in generators.yaml must contain a dict or a list of dicts' % state['path'])
+                continue
+
+            if ordered:
+                case_counter += 1
+
+            for name, value in sorted(case.items(), key=lambda kv: str(kv[0])):
+                if ordered:
+                    num = case_format % case_counter
+                    name = num + ('' if name is None else '-' + str(name))
+                else:
+                    name = str(name)
+
+                next_state = copy.deepcopy(state)
```

Comment:
Doing a deepcopy doesn't feel super nice.

Reply:
Did you have a cleaner alternative in mind?